src/simulators/density_matrix/densitymatrix_state_chunk.hpp

```diff
@@ -0,0 +1,1171 @@
+  uint_t outcome = rng.rand_int(probs);
+  return std::make_pair(outcome, probs[outcome]);
+}
+
+template <class densmat_t>
+void State<densmat_t>::measure_reset_update(const reg_t &qubits,
+                                             const uint_t final_state,
+                                             const uint_t meas_state,
+                                             const double meas_prob) 
+{
+  int_t i;
+
+  // Update a state vector based on an outcome pair [m, p] from
+  // sample_measure_with_prob function, and a desired post-measurement final_state
+  // Single-qubit case
+  if (qubits.size() == 1) {
+    // Diagonal matrix for projecting and renormalizing to measurement outcome
+    cvector_t mdiag(2, 0.);
+    mdiag[meas_state] = 1. / std::sqrt(meas_prob);
+
+#pragma omp parallel for if(BaseState::chunk_omp_parallel_) private(i) 
+    for(i=0;i<BaseState::num_local_chunks_;i++){
+      BaseState::qregs_[i].apply_diagonal_unitary_matrix(qubits, mdiag);
+    }
+
+    // If it doesn't agree with the reset state update
+    if (final_state != meas_state) {
+#pragma omp parallel for if(BaseState::chunk_omp_parallel_) private(i) 
+      for(i=0;i<BaseState::num_local_chunks_;i++){
+        BaseState::qregs_[i].apply_x(qubits[0]);
+      }
+    }
+  }
+  // Multi qubit case
+  else {
+    // Diagonal matrix for projecting and renormalizing to measurement outcome
+    const size_t dim = 1ULL << qubits.size();
+    cvector_t mdiag(dim, 0.);
+    mdiag[meas_state] = 1. / std::sqrt(meas_prob);
+
+#pragma omp parallel for if(BaseState::chunk_omp_parallel_) private(i) 
+    for(i=0;i<BaseState::num_local_chunks_;i++){
+      BaseState::qregs_[i].apply_diagonal_unitary_matrix(qubits, mdiag);
+    }
+
+    // If it doesn't agree with the reset state update
+    // TODO This function could be optimized as a permutation update
+    if (final_state != meas_state) {
+      // build vectorized permutation matrix
+      cvector_t perm(dim * dim, 0.);
+      perm[final_state * dim + meas_state] = 1.;
+      perm[meas_state * dim + final_state] = 1.;
+      for (size_t j=0; j < dim; j++) {
+        if (j != final_state && j != meas_state)
+          perm[j * dim + j] = 1.;
+      }
+      // apply permutation to swap state
+#pragma omp parallel for if(BaseState::chunk_omp_parallel_) private(i) 
+      for(i=0;i<BaseState::num_local_chunks_;i++){
+        BaseState::qregs_[i].apply_unitary_matrix(qubits, perm);
+      }
+    }
+  }
+}
+
+
+//=========================================================================
+// Implementation: Kraus Noise
+//=========================================================================
+
+template <class densmat_t>
+void State<densmat_t>::apply_kraus(const reg_t &qubits,
+                                    const std::vector<cmatrix_t> &kmats) 
+{
+  int_t i;
+  // Convert to Superoperator
+  const auto nrows = kmats[0].GetRows();
+  cmatrix_t superop(nrows * nrows, nrows * nrows);
+  for (const auto kraus : kmats) {
```

Comment:
To avoid copying `kmats` elements:
```suggestion
  for (const auto& kraus : kmats) {
```

Reply:
done